File: app/src/main/java/uk/co/droidinactu/elibrary/library/TagTree.java
package uk.co.droidinactu.booklib.library;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class TagTree {

    Map<Long, Tag> tagMap = new HashMap<>();
    public List<Tag> rootTags = new ArrayList<>();

    public void add(BookTag t) {
        Tag newT = null;
        if (tagMap.containsKey(t.getId())) {
            newT = tagMap.get(t.getId());
        } else {
            newT = new Tag();
            newT.me = t;
            tagMap.put(t.getId(), newT);
        }

        if (newT.me.getParentTagId() == null) {
            rootTags.add(newT);
        } else {
            Tag pt = tagMap.get(newT.me.getParentTagId());
            pt.children.add(newT);
        }
    }

    public void remove(String currentlyReading) {
    }

    public class Tag {
        public BookTag me;
        public List<Tag> children = new ArrayList<>();
    }


}

File: ebooklibrary/src/main/java/uk/co/droidinactu/elibrary/library/TagTree.kt
package uk.co.droidinactu.elibrary.library

import android.app.Dialog
import uk.co.droidinactu.elibrary.room.Tag
import java.util.*

class TagTree {

    private var tagMap: MutableMap<Int, TreeTag> = HashMap()
    var rootTags: MutableList<TreeTag> = ArrayList()

    fun add(t: Tag) {
        var newT: TreeTag?
        if (tagMap.containsKey(t.getUniqueId())) {
            newT = tagMap[t.getUniqueId()]
        } else {
            newT = TreeTag()
            newT.me = t
            tagMap[t.getUniqueId()] = newT
        }

        if (newT!!.me!!.parentTagId == null) {
            rootTags.add(newT)
        } else {
            val pt = tagMap[newT.me!!.parentTagId]
            pt!!.children.add(newT)
        }
    }

    fun remove(currentlyReading: String) {}

    class TreeTag {
        var me: Tag? = null
        var children: ArrayList<TreeTag> = ArrayList()
        var dialog: Dialog? = null
    }

}
